mitmproxy/addons/maplocal.py

```diff
@@ -0,0 +1,135 @@
+import mimetypes
+import re
+import typing
+from pathlib import Path
+
+from werkzeug.security import safe_join
+
+from mitmproxy import ctx, exceptions, flowfilter, http
+from mitmproxy.addons.modifyheaders import parse_spec
+
+
+class MapLocalSpec(typing.NamedTuple):
+    matches: flowfilter.TFilter
+    regex: str
+    local_path: Path
+
+
+def parse_map_local_spec(option: str) -> MapLocalSpec:
+    filter, regex, replacement = parse_spec(option)
+
+    try:
+        re.compile(regex)
+    except re.error as e:
+        raise ValueError(f"Invalid regular expression {regex!r} ({e})")
+
+    try:
+        path = Path(replacement).expanduser().resolve(strict=True)
+    except FileNotFoundError as e:
+        raise ValueError(f"Invalid file path: {replacement} ({e})")
+
+    return MapLocalSpec(filter, regex, path)
+
+
+def _safe_path_join(root: Path, untrusted: str) -> Path:
+    """Join a Path element with an untrusted str.
+
+    This is a convenience wrapper for werkzeug's safe_join,
+    raising a ValueError if the path is malformed."""
+    untrusted_parts = Path(untrusted).parts
+    joined = safe_join(
+        root.as_posix(),
+        *untrusted_parts
+    )
+    if joined is None:
+        raise ValueError("Untrusted paths.")
+    return Path(joined)
+
+
+def file_candidates(url: str, spec: MapLocalSpec) -> typing.List[Path]:
+    """
+    Get all potential file candidates given a URL and a mapping spec ordered by preference.
+    This function already assumes that the spec regex matches the URL.
+    """
+    m = re.search(spec.regex, url)
+    assert m
+    if m.groups():
+        suffix = m.group(1)
+    else:
+        suffix = re.split(spec.regex, url, maxsplit=1)[1]
+        suffix = suffix.split("?")[0]  # remove query string
+
+    suffix = re.sub(r"[^0-9a-zA-Z\-_.=(),/]", "_", suffix.strip("/"))
```

Comment:
URL decode `suffix` here? What exactly is the intent of removing the special characters?

Reply:
This causes

```
spa%20ce.jpg
spa_20ce.jpg
```

instead of `spa ce.jpg`

Reply:
urldecode: yes. Let's fix that and add a test.

Special characters: Some characters are not valid in filenames, hence we want to escape them so that those files can be represented (e.g., `?` on Windows). This very much varies across OSes unfortunately, I think we don't feel like special-casing everything here. One thing we could do is add the unescaped version and the escaped version to our candidate list. What do you think?

Reply:
I see, I didn't think about the case where you want to replace a file that _your_ file system does not support. Without removing them you couldn't do it at all.

> One thing we could do is add the unescaped version and the escaped version (if it differs) to our candidate list.

Yes I like that. The original version should be first in the list which will still allow you to use special characters in 99% of cases (like my two examples). And for this one edge case where there's a character that your filesystem cannot handle then you're still good.